src/main/scala/rise/core/equality.scala

```diff
@@ -198,7 +205,10 @@ object equality {
       case Literal(_: IndexData) => 93
       case Literal(_: ArrayData) => 95
       case Literal(_: PairData) => 97
-      case p: Primitive => p.name.hashCode() + typeEq.hash[TypeKind](p.t)
+      case Opaque(e, t) => 101*hash(e) + 103*typeEq.hash[TypeKind](t)
+      case TypeAnnotation(e, t) => 107*hash(e) + 109*typeEq.hash[TypeKind](t)
+      case TypeAssertion(e, t) => 113*hash(e) + 127*typeEq.hash[TypeKind](t)
+      case p: Primitive => 131*p.name.hashCode() + 137*typeEq.hash[TypeKind](p.t)
```

Comment:
can we use `p.hashCode()`?

Reply:
No, since that would take the type into account.

Reply:
I feel like having the type field straight in our primitive classes keeps preventing us from using default `equals` and `hashCode` implementations, and pushes us to use string names which is inefficient and does not feel like the rigt way. Maybe we should think about extracting the type field from our primitives. What if our expressions looked like:
```scala
sealed trait UntypedExpr
case class UntypedExprNode(node: Node[UntypedExpr])
case class TypeAnnotation(e: UntypedExpr, t: Type) extends UntypedExpr

case class TypedExpr(node: Node[TypedExpr], t: Type)

sealed trait Node[+E]
case class Var(debug: String, id: Int) extends Node[Nothing]
case class App[E](f: E, e: E) extends Node[E]
case class Lambda[E](e: E) extends Node[E]
// ...
abstract class Primitive extends Node[Nothing] {
  def typeScheme: Type
  def name: String
}
```
Then we could use the default `hashCode` and `equals` for primitives when we care only about the primitive, and not about its type, which is not currently the case.

Reply:
That would also solve #152 and #134

Reply:
Maybe I'm missing something, but in that example, if `Primitive <: Node[Nothing]`, would that not mean that in `App(f, e)` if `f : Primitive` then necessarily `e` has to be a subtype of `Node[Nothing]`?

Reply:
I don't think that you will ever need a `Node[Node[_]]` so `f: Primitive` sounds unlikely to be useful. Instead you will have `f: TypedExpr` or `f: UntypedExpr` and then something like `f = TypedExpr(primitive, t)`.

Reply:
You can have a look at my `typed-eqsat` branch to see expressions defined pretty much like above, and how they are used: https://github.com/rise-lang/shine/blob/a25e8b9887974d4b19dce9a4b1b6294f33feaa02/src/main/scala/rise/eqsat/Expr.scala
One of the nice aspects is that you can use that same node definition to define patterns for rewriting which can also include pattern variables on top of the regular nodes (just like we can add type annotations for untyped expressions in the example above), this would be useful if we decide to define our rewrite rule using patterns for Elevate as well.

Reply:
Right, I got confused because I thought `Nothing` was of type `Optional[T]`, when it is in fact of type `Any`. This looks pretty good to me! Should we merge this as is and open a separate PR (it will likely take considerably more work and discussions)